File: app/src/main/java/com/example/newsdemoapplication/ui/home/HomeFragment.java
package com.example.newsdemoapplication.ui.home;

import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.Vibrator;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.HalfFloat;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.newsdemoapplication.NewsVo;
import com.example.newsdemoapplication.R;
import com.example.newsdemoapplication.Util;
import com.example.newsdemoapplication.adapter.ListDragAdapter;
import com.example.newsdemoapplication.adapter.NewsContentAdapter;
import com.example.newsdemoapplication.callback.OnItemClickListener;
import com.example.newsdemoapplication.ui.dashboard.ItemHelper;
import com.example.newsdemoapplication.callback.ItemDragHelperCallBack;
import com.example.newsdemoapplication.ui.login.LoginActivity;
import com.example.newsdemoapplication.util.CommonUtils;
import com.example.newsdemoapplication.view.DragRecycleView;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.lxj.xpopup.XPopup;
import com.lxj.xpopup.interfaces.OnInputConfirmListener;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static com.example.newsdemoapplication.Util.getData;

public class HomeFragment extends Fragment {
    final int SINGLE_LINE =0;
    final int HALF_EXPANDED=1;
    final int COMPLETE_EXPANDED=2;

    final int ColumnNum =4;

    private HomeViewModel homeViewModel;
    ImageView btnExpand;
    ImageView btnFold;
    DragRecycleView mTitleRecycleView;

    GridLayoutManager titleLayoutManager ;
    ListDragAdapter titleAdapter;
    RecyclerView mContentRecycleView;
    BottomSheetDialog bottomSheetDialog;
    //标题栏是否支持局部展开
    private boolean supportHalfExpand=false;
    private int currSelectedPosition;
    private int titleMaxHeight;//像素

    private int currTitleState;
    List<NewsVo> mData =new ArrayList<>();
    List<String> list=new ArrayList<>();
    private int singleLineHeight;
    private NewsContentAdapter mContenAdapter;
    private boolean longPress;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        homeViewModel =
                new ViewModelProvider(this).get(HomeViewModel.class);
        View root = inflater.inflate(R.layout.fragment_home, container, false);
        final TextView textView = root.findViewById(R.id.btn_login);
        homeViewModel.getText().observe(getViewLifecycleOwner(), new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                textView.setText(s);
                textView.setOnClickListener(v -> {
                    startActivity( new Intent(getContext(),LoginActivity.class));
                });
            }
        });
        return root;
    }
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        titleMaxHeight=CommonUtils.dp2px(getActivity(),200);
        singleLineHeight=CommonUtils.dp2px(getActivity(),40);
        list =Util.getData(47);
        initTopContainer();
        initTitleRecycleView();
        initContentRecycleView();
        initBottomSheetDiaLog();
        initLoginBtn();
    }

    private void initLoginBtn() {
    }

    private void initTopContainer() {
        btnExpand =getView().findViewById(R.id.btn_expand);
        btnFold =getView().findViewById(R.id.btn_fold);
        btnFold.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showSingLine();
            }
        });
        btnExpand.setOnClickListener(v->{
            if(currTitleState==COMPLETE_EXPANDED){
                if(supportHalfExpand)
                showHalfExpand();
                else showSingLine();
            }else if(currTitleState==HALF_EXPANDED){
                showCompleteExpand();
            }else {
                if(supportHalfExpand)
                    showHalfExpand();
                else showCompleteExpand();
            }
        });
    }

    private void initTitleRecycleView() {
        mTitleRecycleView =getView().findViewById(R.id.rv_title);
        titleLayoutManager =new GridLayoutManager(requireContext(),ColumnNum*3);
        titleLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                int end=list.size()%ColumnNum;
                if(list.size()-position<=end)
                    return ColumnNum*3/end;
                else
                return 3;
            }
        });
        titleAdapter =new ListDragAdapter(getActivity(),list);
        titleAdapter.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                currSelectedPosition =position;
                titleAdapter.setCurrSelectPosition(position);
                contentScrollToPosition(position);
            }

            @Override
            public void onItemLongClick(View view, int position) {
            }
        });
        mTitleRecycleView.setLayoutManager(titleLayoutManager);
        mTitleRecycleView.setAdapter(titleAdapter);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new ItemDragHelperCallBack(new ItemHelper() {
            @Override
            public void itemMoved(int oldPosition, int newPosition) {
                Util.Loge("move");
                //交换变换位置的集合数据
                Collections.swap(titleAdapter.getData(), oldPosition, newPosition);
                titleAdapter.notifyItemMoved(oldPosition, newPosition);
                currSelectedPosition =newPosition;
//                titleAdapter.setCurrSelectPosition(currSelectedPosition);
            }

            @Override
            public void itemDismiss(int position) {

            }

            @Override
            public void itemClear(int position) {
//                mContenAdapter.notifyDataSetChanged();
//                if(position!=currSelectedPosition){
                    currSelectedPosition=position;
                    titleAdapter.setCurrSelectPosition(currSelectedPosition);
                    titleAdapter.notifyDataSetChanged();
                    if(longPress){
                        bottomSheetDialog.show();
                    }
//                }

            }

            @Override
            public void itemSelected(int position) {//选中title 后的响应操作
                if(position<0) return;
                mTitleRecycleView.setCount(1);
                if(position!=currSelectedPosition){
                    ((ListDragAdapter.MyDragViewHolder)mTitleRecycleView.findViewHolderForAdapterPosition(currSelectedPosition)).setSelected(false);
                    currSelectedPosition =position;
                }
                Vibrator vib = (Vibrator) getContext().getSystemService(Service.VIBRATOR_SERVICE);
                vib.vibrate(150);
//                 记录弹窗信息
                longPress =true;
            }
        }));
        itemTouchHelper.attachToRecyclerView(mTitleRecycleView);
        mTitleRecycleView.setMCallBack(new DragRecycleView.DragAndPressCallBack() {//设置移动监听回调
            @Override
            public void onLongPress() {
                bottomSheetDialog.show();
                mTitleRecycleView.setBottomShowTime(System.currentTimeMillis());
            }
            @Override
            public void onAfterPressMove() {
                longPress=false;
                bottomSheetDialog.dismiss();
            }
        });

        mTitleRecycleView.postDelayed((Runnable) () -> { //视图刷新后，计算视图高度
            Util.Loge(String.format("height %d",mTitleRecycleView.getMeasuredHeight()));
            if(mTitleRecycleView.getMeasuredHeight()>titleMaxHeight){
                supportHalfExpand =true;
                showHalfExpand();
            }else {
                currTitleState =COMPLETE_EXPANDED;
            }
        },50);
    }

    private void contentScrollToPosition(int position) {
        LinearLayoutManager mLayoutManager =
                (LinearLayoutManager) mContentRecycleView.getLayoutManager();
        mLayoutManager.scrollToPositionWithOffset(position,0);
//        mContentRecycleView.offsetTopAndBottom(0);
    }

    private void showSingLine() {
        currTitleState = SINGLE_LINE;
        LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) mTitleRecycleView.getLayoutParams();
        layoutParams.height =singleLineHeight;
        mTitleRecycleView.setLayoutParams(layoutParams);

        showExpandFoldBtns();
    }

    private void showHalfExpand() {
        currTitleState = HALF_EXPANDED;
        LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) mTitleRecycleView.getLayoutParams();
        layoutParams.height =titleMaxHeight;
        mTitleRecycleView.setLayoutParams(layoutParams);

        showExpandFoldBtns();
    }
    private void showCompleteExpand() {
        currTitleState = COMPLETE_EXPANDED;
        LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) mTitleRecycleView.getLayoutParams();
        layoutParams.height =LinearLayout.LayoutParams.WRAP_CONTENT;
        mTitleRecycleView.setLayoutParams(layoutParams);
        showExpandFoldBtns();
    }

    /**
     * 根据标题栏折叠状态显示按钮
     */
    private void showExpandFoldBtns() {
        switch (currTitleState){
            case SINGLE_LINE:
                btnExpand.setImageResource(R.drawable.ic_baseline_expand_more_24);
                btnFold.setVisibility(View.GONE);
                break;
            case HALF_EXPANDED:
                btnExpand.setImageResource(R.drawable.ic_baseline_expand_more_24);
                btnFold.setVisibility(View.VISIBLE);
                break;
            case COMPLETE_EXPANDED:
                btnExpand.setImageResource(R.drawable.ic_baseline_expand_less_24);
                btnFold.setVisibility(View.VISIBLE);
                break;
            default:break;
        }
    }

    private void initContentRecycleView() {
        mContentRecycleView = getView().findViewById(R.id.rv_content_list);
        mContenAdapter =new NewsContentAdapter(getContext(),list);
        mContentRecycleView.setAdapter(mContenAdapter);
        mContentRecycleView.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            LinearLayoutManager linearLayoutManager = (LinearLayoutManager) mContentRecycleView.getLayoutManager();
            int topPosition =linearLayoutManager.findFirstVisibleItemPosition();
            if(topPosition>=0&&topPosition!=currSelectedPosition){
                currSelectedPosition=topPosition;
                titleAdapter.setCurrSelectPosition(currSelectedPosition);
                Log.e("TAG", "initContentRecycleView: "+ topPosition);
                titleLayoutManager.scrollToPositionWithOffset(topPosition,0);
            }
        });
    }
    private void initBottomSheetDiaLog() {
        bottomSheetDialog =new BottomSheetDialog(getContext());
        bottomSheetDialog.setContentView(R.layout.botton_dialog);
        bottomSheetDialog.setCancelable(true);
        bottomSheetDialog.findViewById(R.id.tv_cancel).setOnClickListener(v -> {
            bottomSheetDialog.dismiss(); }
        );
        bottomSheetDialog.findViewById(R.id.tv_rename).setOnClickListener(v -> {
            Log.e("TAG", "initBottomSheetDiaLog: "+currSelectedPosition );
            bottomSheetDialog.dismiss();
                    new XPopup.Builder(getContext()).asInputConfirm("重命名", "请输入内容。",
                            text -> {
                                list.set(currSelectedPosition,text);
                                titleAdapter.notifyItemChanged(currSelectedPosition);
                                mContenAdapter.notifyItemChanged(currSelectedPosition);
                            })
                            .show();
        }
        );
        bottomSheetDialog.findViewById(R.id.tv_delete).setOnClickListener(v -> {
            Log.e("TAG", "initBottomSheetDiaLog: "+currSelectedPosition );
            list.remove(currSelectedPosition);
            titleAdapter.notifyItemRemoved(currSelectedPosition);
            mContenAdapter.notifyItemRemoved(currSelectedPosition);
            bottomSheetDialog.dismiss(); }
        );
    }

}

File: app/src/main/java/com/example/newsdemoapplication/ui/home/HomeFragment.kt
package com.example.newsdemoapplication.ui.home

import android.annotation.SuppressLint
import android.app.Service
import android.os.Bundle
import android.os.Vibrator
import android.util.Log
import android.view.View
import android.widget.LinearLayout
import androidx.core.view.GravityCompat
import androidx.lifecycle.observe
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.example.newsdemoapplication.Constants
import com.example.newsdemoapplication.R
import com.example.newsdemoapplication.databinding.HomeFragmentBinding
import com.example.newsdemoapplication.view.ListDragAdapter
import com.example.newsdemoapplication.view.ListDragAdapter.MyDragViewHolder
import com.example.newsdemoapplication.util.callback.ItemHelper
import com.example.newsdemoapplication.util.CommonUtils
import com.example.newsdemoapplication.util.callback.ItemDragHelperCallBack
import com.example.newsdemoapplication.util.callback.OnItemClickListener
import com.example.newsdemoapplication.view.ChapterDragView
import com.example.newsdemoapplication.view.TitleRecycleView.DragAndPressCallBack
import com.example.newsdemoapplication.model.ChapterVo
import com.example.newsdemoapplication.ui.BlackFragment
import com.example.newsdemoapplication.ui.main_activity.MainActivity
import com.example.newsdemoapplication.ui.manager.AddSectionFragment
import com.example.newsdemoapplication.util.callback.SimpleCallback
import com.example.newsdemoapplication.util.common.MvvmBaseFragment
import com.example.newsdemoapplication.util.getRandomData
import com.example.newsdemoapplication.util.log
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.lxj.xpopup.XPopup
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.internal.notify
import java.util.*

/**
 * app首页
 * @author gzp
 */
class HomeFragment : MvvmBaseFragment<HomeViewModel, HomeFragmentBinding>(),CoroutineScope by MainScope(),SimpleCallback {
    private val TAG =this::class.java.simpleName
    override fun getLayoutResId()=R.layout.home_fragment
    val mainActivity by lazy { activity as MainActivity }
    //标题栏最大列数
    val ColumnNum = 4
    //顶部标题栏折叠状态
    private val SINGLE_LINE = 0
    private val HALF_EXPANDED = 1
    private val COMPLETE_EXPANDED = 2

    private var currTitleState = SINGLE_LINE


    private var editChapter: Boolean=false
    val chapterDragView:ChapterDragView by lazy { mainActivity.binding.zuoyeId.listView.also {
        it.mAdapter.selectCallback=this
        it.mCallBack = object : DragAndPressCallBack {
            //设置移动监听回调
            override fun onLongPress() {
                bottomSheetDialog.show()
                mTitleRecycleView.bottomShowTime = System.currentTimeMillis()
            }

            override fun onAfterPressMove() {
                longPress = false
                bottomSheetDialog.dismiss()
            }
        }
    }}

    var currContentNewsItem=0
    lateinit var mFragmentAdapter:FragmentStateAdapter
    //回到页面时更新数据
    override fun onResume() {
        super.onResume()
        log("resume")
        arguments?.getSerializable(Constants.ChapterVo)?.let {
            vm.saveChapter(it as ChapterVo)
        }
    }

    override fun doCreateView(savedInstanceState: Bundle?) {
        btnFold //初始化 首页顶部标题栏 收起按钮
        btnExpand //初始化 首页顶部标题栏 向下展开按钮
        initView()
        initTitleRecycleView()
        initObservable()
        launch {
            delay(500)
            vm.randomData()
        }
    }


    /**
     * 设置页面点击响应事件
     */
    @SuppressLint("ClickableViewAccessibility")
    private fun initView() {
        mainActivity.binding.zuoyeId.tjanjiadidianAn.setOnClickListener {
            mainActivity.closeDrawerLayout()
            start(AddSectionFragment())
        }
        binding.apply {
            mFragmentAdapter=object :FragmentStateAdapter(this@HomeFragment){
                override fun getItemCount()=currContentNewsItem
                override fun createFragment(position: Int): ScrollSliderFragment {
                    val fragment = ScrollSliderFragment()
                    fragment.arguments = Bundle().apply {
                        putInt(Constants.SLIDE_OBJECT, position )
                    }
                    fragment.newsVo=vm.findNewsVo(position)
                    return fragment//.also { it.update(vm.currChapter.value?.listNews,position) }
                }
            }
            pager.adapter=mFragmentAdapter
            pager.registerOnPageChangeCallback(object :ViewPager2.OnPageChangeCallback(){
                override fun onPageSelected(position: Int) {
                    super.onPageSelected(position)
                    log("currSelectPosition ${titleAdapter.currSelectPosition}滚到 $position")
                    if (titleAdapter.currSelectPosition != position) {
                        titleAdapter.currSelectPosition=position
                        titleLayoutManager.scrollToPositionWithOffset(position,0)
                    }
                }
            })
            (activity as? MainActivity)?.let {a->
                llTopLeft.setOnClickListener {// 左侧弹窗按钮
                    a.mDrawerLayout.openDrawer(GravityCompat.START) }
                    chapterDragView.mAdapter.notifyDataSetChanged()
                llTopRight.setOnClickListener {
                    a.mDrawerLayout.openDrawer(GravityCompat.END)
                }
            }
            xntyTextViewId.setOnClickListener {  log("生成假数据")
                vm.randomData() }
        }

    }

    //监控到关键数据变化时的UI响应逻辑
    private fun initObservable() {
        vm.apply {
            listChapter.observe(this@HomeFragment){
                chapterDragView.mAdapter.list=it
                chapterDragView.mAdapter.notifyDataSetChanged()

                toastShort(it.size.toString())
            }
            currChapter.observe(this@HomeFragment){
                log("数据初始化")
                contentScrollToPosition(0)

                binding.apply {
                    suoTextViewId.text=it.getLockStr()
                    ddTextViewId.text=it.chapterName
                    txTextViewId.text=it.description
                }
                titleAdapter.setmDatas(it.listNews?.map { it.title }?.apply {
                    if(size<=ColumnNum){
                        btnFold.visibility=View.GONE
                        btnExpand.visibility=View.GONE
                    }else{
                        btnExpand.visibility=View.VISIBLE
                    }
                })
//                mContentAdapter.setData(it.listNews)
                mFragmentAdapter.notifyDataSetChanged()
            }
            toast.observe(this@HomeFragment){
                toastShort(it)
            }
        }
    }


    //顶部 抽屉展开按钮
    val btnExpand by lazy { binding.btnExpand.apply {
        Log.e(TAG, ": 抽屉展开按钮", )
        setOnClickListener {
//            if(vm.currChapter.value?.listNews?.size?:0<=ColumnNum){
//                log("没有更多了")
//                return@setOnClickListener
//            }
            if (currTitleState == COMPLETE_EXPANDED) {
                if (supportHalfExpand) updateTitleButtons(HALF_EXPANDED)
                else updateTitleButtons()
            } else if (currTitleState == HALF_EXPANDED) {
                updateTitleButtons(COMPLETE_EXPANDED)
            } else {
                if (supportHalfExpand) updateTitleButtons(HALF_EXPANDED)
                else updateTitleButtons(COMPLETE_EXPANDED)
            }
        }
    } }
    //顶部标题列表栏视图
    private val mTitleRecycleView by lazy {
        binding.rvTitle.apply {
            layoutManager = titleLayoutManager
            adapter = titleAdapter
        }
    }
    //标题栏底部折叠按钮
    val btnFold by lazy {
        Log.e(TAG, ": 合上展开按钮", )
        binding.btnFold.also {
            it.setOnClickListener { updateTitleButtons() }
        }
    }

    //顶部标题栏布局管理
    private val titleLayoutManager by lazy {
        GridLayoutManager(requireContext(), ColumnNum * 3).apply {
            spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    val end: Int = listData.size % ColumnNum
                    return if (listData.size - position <= end) ColumnNum * 3 / end else 3
                }
            }
        }
    }
    //顶部标题栏适配器
    val titleAdapter by lazy {
        ListDragAdapter(activity, listData).apply {
            setOnItemClickListener(
                    object : OnItemClickListener {
                        override fun onItemClick(view: View, position: Int) {
                            currSelectedPosition = position
                            currSelectPosition = position
                            contentScrollToPosition(position)
                        }
                        override fun onItemLongClick(view: View, position: Int) {}
                    })
        }
    }
    //底部弹窗设置
    val bottomSheetDialog by lazy { BottomSheetDialog(requireContext()).apply {
        setContentView(R.layout.bottom_dialog_style)
        setCancelable(true)
        findViewById<View>(R.id.tv_cancel)?.setOnClickListener {dismiss() }
        findViewById<View>(R.id.tv_rename)?.setOnClickListener {
            Log.e("TAG", "initBottomSheetDiaLog: $currSelectedPosition")
            dismiss()
//            val leftListOpen =mainActivity.mDrawerLayout.isDrawerOpen(mainActivity.binding.zuochoutiId)
            val leftListOpen =true
            if(editChapter&&! leftListOpen){
                editChapter=false //todo 跳转到编辑页
            }else
            XPopup.Builder(activity).asInputConfirm("重命名 $leftListOpen", "请输入内容。") { text: String? ->
                if(!leftListOpen){
                    listData[currSelectedPosition] = text?:"-"
                    titleAdapter.notifyItemChanged(currSelectedPosition)
                    mFragmentAdapter.notifyItemChanged(currSelectedPosition)
//                                        mContentAdapter.notifyItemChanged(currSelectedPosition)
                }else{
                    vm.currChapter.value?.let {
                        val a =it
                        a.chapterName=text?:"-"
                        vm.listChapter.value?.firstOrNull { it.id==a.id }?.let { it.chapterName=a.chapterName }
                        vm.currChapter.postValue(a)
                        vm.listChapter.value=vm.listChapter.value
                    }

                }
            }.show()
        }
        findViewById<View>(R.id.tv_delete)?.setOnClickListener { v: View? ->
//            val leftListOpen =mainActivity.mDrawerLayout.isDrawerOpen(mainActivity.binding.zuochoutiId)
            val leftListOpen =true
            if(editChapter&&! leftListOpen){
                editChapter=false
                vm.listChapter.apply {
                    val index =value?.indexOf(vm.currChapter.value)
                    index?.takeIf { it>=0 }?.let {
                        val list =value?.also { it.removeAt(index ?: 0) }
                        postValue(list)
                    }
                }
            }else{
                Log.e("TAG", "initBottomSheetDiaLog: $currSelectedPosition")

                if(leftListOpen){
                    vm.deleteChapter()
//                    mainActivity.mDrawerLayout.closeDrawer(mainActivity.binding.zuochoutiId)
                }else{
                    listData.removeAt(currSelectedPosition)
                    titleAdapter.notifyItemRemoved(currSelectedPosition)
                    mFragmentAdapter.notifyItemRemoved(currSelectedPosition)
                }

            }
            dismiss()

        }
    } }

    //标题栏是否支持局部展开
    private var supportHalfExpand = false
    private var currSelectedPosition = 0
    private var titleMaxHeight //像素
            = CommonUtils.dp2px(activity, 200f)
    private val singleLineHeight by lazy {
        CommonUtils.dp2px(activity, 40f)
    }


    val listData= getRandomData(17)

    private var longPress = false


    //初始化上部小标题列表栏
    private fun initTitleRecycleView() {
        val itemTouchHelper = ItemTouchHelper(ItemDragHelperCallBack(object : ItemHelper {
            override fun itemMoved(oldPosition: Int, newPosition: Int) {
                log("move")
                //交换变换位置的集合数据
                Collections.swap(titleAdapter.data, oldPosition, newPosition)
                titleAdapter.notifyItemMoved(oldPosition, newPosition)
                currSelectedPosition = newPosition
            }

            override fun itemDismiss(position: Int) {}
            override fun itemClear(position: Int) {
                currSelectedPosition = position
                titleAdapter.currSelectPosition = currSelectedPosition
                titleAdapter.notifyDataSetChanged()
                if (longPress) {
                    bottomSheetDialog.show()
                    longPress = false
                }
            }

            override fun itemSelected(position: Int) { //选中title 后的响应操作
                if (position < 0) return
                mTitleRecycleView.count = 2
                if (position != currSelectedPosition) {
                    (mTitleRecycleView.findViewHolderForAdapterPosition(currSelectedPosition) as? MyDragViewHolder)?.setSelected(false)
                    currSelectedPosition = position
                }
                val vib = context!!.getSystemService(Service.VIBRATOR_SERVICE) as Vibrator
                vib.vibrate(150)
                //                 记录弹窗信息
                longPress = true
            }
        }))
        itemTouchHelper.attachToRecyclerView(mTitleRecycleView)
        mTitleRecycleView.mCallBack = object : DragAndPressCallBack {
            //设置移动监听回调
            override fun onLongPress() {
                bottomSheetDialog.show()
                mTitleRecycleView.bottomShowTime = System.currentTimeMillis()
            }

            override fun onAfterPressMove() {
                longPress = false
                bottomSheetDialog.dismiss()
            }
        }
        mTitleRecycleView.postDelayed({
            //视图刷新后，计算视图高度
            log(String.format("height %d", mTitleRecycleView.measuredHeight))
            if (mTitleRecycleView.measuredHeight > titleMaxHeight) {
                supportHalfExpand = true
                updateTitleButtons(HALF_EXPANDED)
            } else {
                currTitleState = COMPLETE_EXPANDED
            }
        }, 50)
    }

    private fun contentScrollToPosition(position: Int) {
       vm.currChapter.value?.listNews?.let {
           currContentNewsItem=it.size
       }

        log("${vm.currChapter.value}itemCount ${mFragmentAdapter.itemCount} 页")
        binding.pager.setCurrentItem(position,false)
        mFragmentAdapter.createFragment(position)
    }


    /**
     * 根据顶部标题栏折叠状态，显示标题栏按钮
     * 刷新标题栏按钮显示
     */
    private fun updateTitleButtons(newState:Int=SINGLE_LINE) {
        currTitleState=newState
        val layoutParams = mTitleRecycleView.layoutParams as LinearLayout.LayoutParams
        when (newState) {
            SINGLE_LINE -> {
                layoutParams.height = singleLineHeight
                btnExpand.setImageResource(R.drawable.ic_baseline_expand_more_24)
                btnFold.visibility = View.GONE
            }
            HALF_EXPANDED -> {
                layoutParams.height = titleMaxHeight
                btnExpand.setImageResource(R.drawable.ic_baseline_expand_more_24)
                btnFold.visibility = View.VISIBLE
            }
            COMPLETE_EXPANDED -> {
                layoutParams.height = LinearLayout.LayoutParams.WRAP_CONTENT
                btnExpand.setImageResource(R.drawable.ic_baseline_expand_less_24)
                btnFold.visibility = View.VISIBLE
            }
            else -> {
            }
        }
        mTitleRecycleView.layoutParams = layoutParams
    }

    fun refreshDate() {
        vm.randomData()
    }

    override fun changeChapter(id: Int) {
        log("changeChapter $id")
        vm.changeChapter(id)
    }
}